src/Arcus.WebApi.Security/Authentication/CertificateAuthenticationValidator.cs

```diff
@@ -0,0 +1,252 @@
+            return this;
+        }
+
+        /// <summary>
+        /// Validate the specified <paramref name="clientCertificate"/> based on the given <paramref name="requirements"/>
+        /// using the <paramref name="services"/> to retrieve registered instances that will provide the expected certificate values.
+        /// </summary>
+        /// <param name="clientCertificate">The client certificate to validate.</param>
+        /// <param name="requirements">The requirements to identify which parts of the <paramref name="clientCertificate"/> should be validated.</param>
+        /// <param name="services">The collection of registered services, (ex. from the request pipeline).</param>
+        /// <returns>
+        ///     <c>true</c> when the specified <paramref name="clientCertificate"/> is valid, <c>false</c> otherwise.
+        /// </returns>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="clientCertificate"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="requirements"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="services"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentException">Thrown when the <paramref name="requirements"/> contains a configuration key that is <c>null</c>.</exception>
+        /// <exception cref="KeyNotFoundException">
+        ///     Thrown when the <paramref name="requirements"/> requires to validate a property of the <paramref name="clientCertificate"/>
+        ///     for which there exists no registered <see cref="IX509ValidationLocation"/> implementation.
+        /// </exception>
+        internal async Task<bool> ValidateCertificate(
+            X509Certificate2 clientCertificate,
+            IDictionary<X509ValidationRequirement, ConfiguredKey> requirements,
+            IServiceProvider services)
+        {
+            Guard.NotNull(clientCertificate, nameof(clientCertificate), "Certificate authentication validation requires a client certificate");
+            Guard.NotNull(requirements, nameof(requirements), "Certificate authentication validation requires a series of requirements and their configured key");
+            Guard.NotNull(services, nameof(services), "Certificate authentication validation requires a service object to retrieve registered services");
+            Guard.For<ArgumentException>(
+                () => requirements.Any(requirement => requirement.Value is null), 
+                "Certificate authentication requires all configured keys in the series of requirements to be not 'null'");
+
+            IDictionary<X509ValidationRequirement, (ConfiguredKey configuredKey, string expected)> requirementValues = 
+                await GetExpectedCertificateValues(requirements, services);
+            
+            ILogger logger = 
```

Comment:
Same feedback, do logger determination here or move to small method?

Reply:
I restructured some of the methods so it requires less arguments and is more simpler.

Reply:
THanks!